lib/puppet/provider/package/pkgng.rb

```diff
@@ -61,7 +61,7 @@ def self.instances
   end
 
   def self.prefetch(resources)
-    packages = instances
+    packages = instances || []
```

Comment:
I'm not sure that hiding such a failure is the correct behavior.

I can't find other package providers that work like this.

For instance, the `yum` provider will bubble up errors like this: https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider/package/yum.rb#L28-L31

It relies on the behavior of the `rpm` provider, which also will throw errors if the list of packages cannot be retrieved with a call to `instances` - see https://github.com/puppetlabs/puppet/blob/master/lib/puppet/provider/package/rpm.rb#L78-L95

You've noted in the ticket:


> The fix here is to return an empty array of package instances in the event of a non-zero exit call to pkg(8) during instances enumeration of the existing packages.  This might happen for a variety of reasons, one of which would be an empty list of repositories.


But that's not quite what you've done here.  I think this PR should change the implementation of `instances`, so that when there are no sources it returns an empty array.  If there are other valid reasons for no packages to be listed, we should handle that scenario appropriately.  But if there are complete failures to  execute the command to list packages, then that should be a failure condition with the resource.